src/utils/TraverseState.ts

```diff
@@ -143,17 +143,7 @@ export default class TraverseState {
     if (!this.parentState) {
       return true;
     }
-
-    switch (this.scope.block.type) {
-      case 'FunctionDeclaration':
-      case 'FunctionExpression':
-      case 'ObjectMethod':
-      case 'ClassMethod':
-        return true;
-
-      default:
-        return false;
-    }
+    return isFunction(this.scope.block);
```

Comment:
Does this already handle arrow functions that don’t have block bodies?

Reply:
Good point. I thought that it would work, (it's still an ArrowFunctionExpression node) but from my testing now it looks like the variable gets incorrectly hoisted in that case, so something is going wrong. I'll work on a fix.

Reply:
Actually, yeah, this is pretty tricky. This CoffeeScript code should print 2:
```
f = => a = 1
a = 2
f()
console.log a
```

but decaffeinate produces the `a = 1` inside a blockless arrow function. I think the only way add-variable-declarations could correctly handle this is by converting the arrow function into one with a block body so it could add an inner `var a;` statement, which is a lot more complicated than any patching that add-variable-declarations does at the moment.

I think a better approach (at least an easier one) is to change decaffeinate to never produce blockless arrow functions that contain an assignment node. That way, add-variable-declarations will always have a block to add the declaration to. It's sort of a weird interaction between the two projects, but maybe defensible in that add-variable-declarations is copying CoffeeScript's behavior, and CoffeeScript doesn't have blockless arrow functions, so add-variable-declarations shouldn't have an opinion on how to handle them and decaffeinate shouldn't produce code that runs into that issue.

So my current plan is to move forward with this change and then do a separate decaffeinate change to avoid assignments in blockless arrow functions.

Reply:
Sounds good.